In [8]:
import pandas as pd
import numpy as np 

In [9]:
llama_data= pd.read_csv("Llama/llama_resp.csv")
llama_data                                                           

day  meal_type                                          meal_name  \
0      Monday  breakfast             Italian-Style Omelette with Vegetables   
1      Monday      lunch       Grilled Chicken Panini with Tomato and Basil   
2      Monday     dinner  Spaghetti Squash with Meatballs and Marinara S...   
3     Tuesday  breakfast              Greek Yogurt with Berries and Granola   
4     Tuesday      lunch          Turkey and Avocado Wrap with Mixed Greens   
5     Tuesday     dinner                     Baked Cod with Lemon and Herbs   
6   Wednesday  breakfast  Whole-Grain Pancakes with Fresh Fruits and Yogurt   
7   Wednesday      lunch    Veggie and Bean Chili with Whole-Grain Crackers   
8   Wednesday     dinner      Shrimp and Vegetable Stir-Fry with Brown Rice   
9    Thursday  breakfast    Egg and Spinach Omelette with Whole-Grain Toast   
10   Thursday      lunch          Chicken and Vegetable Skewers with Quinoa   
11   Thursday     dinner  Beef and Mushroom Stroganoff with Whole-Grain ...   
12     Friday  breakfast      Cottage Cheese with Fresh Fruits and Cinnamon   
13     Friday      lunch  Grilled Chicken Breast with Roasted Vegetables...   
14     Friday     dinner             Pan-Seared Salmon with Lemon and Herbs   
15   Saturday  breakfast  Whole-Grain Waffles with Fresh Berries and Yogurt   
16   Saturday      lunch  Turkey Meatball Sub with Marinara Sauce and Wh...   
17   Saturday     dinner  Grilled Chicken Breast with Roasted Sweet Pota...   
18     Sunday  breakfast  Breakfast Burrito with Scrambled Eggs and Blac...   
19     Sunday      lunch     Chicken Caesar Salad with Whole-Grain Croutons   
20     Sunday     dinner      Beef and Mushroom Stroganoff with Egg Noodles   

                                          description  calories  protein  \
0   An Italian-style omelette made with eggs, vege...       250       20   
1   A grilled chicken breast topped with tomato, b...       320       35   
2   A baked spaghetti squash topped with homemade ...       400       30   
3   A bowl of Greek yogurt topped with mixed berri...       200       15   
4   A whole-grain wrap filled with sliced turkey b...       350       30   
5   A baked cod fillet seasoned with lemon juice a...       240       25   
6   Fluffy whole-grain pancakes topped with fresh ...       280       10   
7   A hearty chili made with vegetables, beans, an...       400       20   
8   A stir-fry made with shrimp, mixed vegetables,...       320       25   
9   An egg omelette made with spinach, eggs, and w...       220       15   
10  Grilled chicken breast and mixed vegetables se...       380       35   
11  A beef and mushroom stroganoff made with whole...       420       30   
12  A bowl of cottage cheese topped with fresh fru...       200       20   
13  A grilled chicken breast served with roasted v...       300       35   
14  A pan-seared salmon fillet seasoned with lemon...       240       25   
15  A crispy whole-grain waffle topped with fresh ...       260       10   
16  A turkey meatball sub made with marinara sauce...       400       30   
17  A grilled chicken breast served with roasted s...       320       35   
18  A breakfast burrito made with scrambled eggs, ...       240       15   
19  A chicken Caesar salad made with whole-grain c...       340       30   
20  A beef and mushroom stroganoff made with egg n...       420       30   

    carbohydrates  fiber  fat  sugar  is_alternative  
0              10      2   15      5           False  
1              20      4   15      5           False  
2              30      5   15     10           False  
3              25      4    0     20           False  
4              25      5   15     10           False  
5               0      0   10      0            True  
6              40      4    5     20           False  
7              40     10   15     10           False  
8              30      5   10     10           False  
9              10      2   10      5           F

In [10]:
gemini_data = pd.read_csv("Gemini/data_Gemini.csv")
gemini_data

day  meal_type                                          meal_name  \
0      Monday  Breakfast                     Chia Seed Pudding with Berries   
1      Monday  Breakfast                        Scrambled Eggs with Spinach   
2      Monday      Lunch                 Caprese Salad with Grilled Chicken   
3      Monday      Lunch                         Mediterranean Quinoa Salad   
4      Monday     Dinner              Baked Salmon with Asparagus and Lemon   
5      Monday     Dinner                   Chicken Stir-fry with Brown Rice   
6     Tuesday  Breakfast             Greek Yogurt with Almonds and Cinnamon   
7     Tuesday  Breakfast                      Oatmeal with Berries and Nuts   
8     Tuesday      Lunch                                        Lentil Soup   
9     Tuesday      Lunch        Chicken Salad Sandwich on Whole Wheat Bread   
10    Tuesday     Dinner             Zucchini Noodles with Pesto and Shrimp   
11    Tuesday     Dinner  Turkey Meatballs with Marinara Sauce and Zucchini   
12  Wednesday  Breakfast                          Cottage Cheese with Fruit   
13  Wednesday  Breakfast  Smoothie with Spinach, Berries, and Protein Po...   
14  Wednesday      Lunch                      Chicken and Vegetable Skewers   
15  Wednesday      Lunch                           Tuna Salad Lettuce Wraps   
16  Wednesday     Dinner        Eggplant Parmesan (using whole wheat pasta)   
17  Wednesday     Dinner              Chicken Piccata with Zucchini Noodles   
18   Thursday  Breakfast                     Whole Wheat Toast with Avocado   
19   Thursday  Breakfast         Breakfast Burrito with Eggs and Vegetables   
20   Thursday      Lunch                                      Caprese Salad   
21   Thursday      Lunch                Shrimp Scampi with Zucchini Noodles   
22   Thursday     Dinner             Chicken Breast with Roasted Vegetables   
23   Thursday     Dinner         Turkey Meatloaf with Mashed Sweet Potatoes   
24     Friday  Breakfast                 Ricotta Cheese Omelet with Spinach   
25     Friday  Breakfast       Egg White Scramble with Mushrooms and Onions   
26     Friday      Lunch                          Salad with Grilled Salmon   
27     Friday      Lunch                    Leftover Chicken and Vegetables   
28     Friday     Dinner  Spaghetti Squash with Marinara Sauce and Turke...   
29     Friday     Dinner         Baked Cod with Roasted Broccoli and Quinoa   
30   Saturday  Breakfast                      Protein Pancakes with Berries   
31   Saturday  Breakfast              Scrambled Eggs with Whole Wheat Toast   
32   Saturday      Lunch        Chicken Caesar Salad (using light dressing)   
33   Saturday      Lunch                          Leftover Spaghetti Squash   
34   Saturday     Dinner        Pizza with Whole Wheat Crust and Vegetables   
35   Saturday     Dinner      Shrimp and Vegetable Stir-fry with Brown Rice   
36     Sunday  Breakfast             Baked Oatmeal with Apples and Cinnamon   
37     Sunday  Breakfast            Yogurt Parfait with Granola and Berries   
38     Sunday      Lunch                                     Leftover Pizza   
39     Sunday      Lunch               Tuna Salad with Whole Wheat Crackers   
40     Sunday     Dinner  Roasted Chicken with Sweet Potato and Green Beans   
41     Sunday     Dinner                              Lentil Shepherd's Pie   

                                          description  calories  protein  \
0   Chia seeds soaked in almond milk with mixed be...       150        5   
1           Two scrambled eggs with a cup of spinach.       180       15   
2   Fresh mozzarella, tomatoes, basil, and grilled...       250       25   
3   Quinoa, cucumber, tomatoes, olives, feta chees...       220       12   
4   Baked salmon fillet with roasted asparagus and...       300       30   
5   Chicken stir-fry with brown rice and mixed veg...       350       28   
6   Plain Greek yogurt topped with sliced almonds ...       170       15   
7   

In [11]:
avg_cols = ['calories','protein','carbohydrates','fiber','fat','sugar']

In [16]:
gemini_avg = gemini_data[avg_cols].mean()
gemini_avg

calories         249.761905
protein           19.333333
carbohydrates     21.238095
fiber              5.000000
fat                8.642857
sugar              4.928571
dtype: float64

In [17]:
llama_avg = llama_data[avg_cols].mean()
llama_avg

calories         309.523810
protein           24.761905
carbohydrates     23.809524
fiber              4.142857
fat               11.904762
sugar             10.238095
dtype: float64

### Recommended Nutritional Content per 100 g for Meals in Type 2 Diabetes Patients

| Nutrient      | Recommended Range*                 |
|---------------|------------------------------------|
| **Calories**  | 100–200 kcal                      |
| **Protein**   | 5–10 g                            |
| **Carbohydrates** | 10–15 g (preferably low GI carbs) |
| **Fiber**     | 3–6 g                             |
| **Fat**       | 2–6 g (focus on unsaturated fats) |
| **Sugar**     | < 5 g (preferably natural sugars) |

*These values are general guidelines and should be adapted to individual needs under a dietitian's supervision.

### Sources
1. UCSF Diabetes Education Online [UCSF](https://dtc.ucsf.edu)  
2. Harvard T.H. Chan School of Public Health [Harvard Health](https://www.hsph.harvard.edu/nutritionsource)  
3. American Diabetes Association recommendations [ADA](https://www.diabetes.org).  

Always consult a healthcare provider for personalized advice.

In [18]:
requirements_data = {
    "Calories (kcal)": "50-150",
    "Protein (g)": "4-10",
    "Carbohydrates (g)": "10-20",
    "Fiber (g)": "2-6",
    "Fat (g)": "1-8",
    "Sugar (g)": "<5"
}

In [19]:
requirements_data

{'Calories (kcal)': '50-150',
 'Protein (g)': '4-10',
 'Carbohydrates (g)': '10-20',
 'Fiber (g)': '2-6',
 'Fat (g)': '1-8',
 'Sugar (g)': '<5'}

In [20]:
# Define the real requirements as midpoints of the ranges
requirements_data = {
    "Calories (kcal)": (50 + 150) / 2,  # midpoint of range
    "Protein (g)": (4 + 10) / 2,
    "Carbohydrates (g)": (10 + 20) / 2,
    "Fiber (g)": (2 + 6) / 2,
    "Fat (g)": (1 + 8) / 2,
    "Sugar (g)": 5  # upper bound used
}

# AI model averages
llama_avg = {
    "Calories (kcal)": 309.52,
    "Protein (g)": 24.76,
    "Carbohydrates (g)": 23.81,
    "Fiber (g)": 4.14,
    "Fat (g)": 11.90,
    "Sugar (g)": 10.24
}

gemini_avg = {
    "Calories (kcal)": 249.76,
    "Protein (g)": 19.33,
    "Carbohydrates (g)": 21.24,
    "Fiber (g)": 5.00,
    "Fat (g)": 8.64,
    "Sugar (g)": 4.93
}

# Convert to DataFrame for easier calculations
real = pd.Series(requirements_data)
llama = pd.Series(llama_avg)
gemini = pd.Series(gemini_avg)

# Calculate metrics
def calculate_metrics(predicted, real):
    absolute_error = np.abs(predicted - real)
    percentage_error = (absolute_error / real) * 100
    mse = np.mean((predicted - real) ** 2)
    rmse = np.sqrt(mse)
    mae = np.mean(absolute_error)
    return pd.DataFrame({
        "Absolute Error": absolute_error,
        "Percentage Error (%)": percentage_error,
        "MSE": mse,
        "RMSE": rmse,
        "MAE": mae
    })

# Metrics for LLaMA and Gemini
llama_metrics = calculate_metrics(llama, real)
gemini_metrics = calculate_metrics(gemini, real)

llama_metrics, gemini_metrics


(                   Absolute Error  Percentage Error (%)          MSE  \
 Calories (kcal)            209.52            209.520000  7395.650217   
 Protein (g)                 17.76            253.714286  7395.650217   
 Carbohydrates (g)            8.81             58.733333  7395.650217   
 Fiber (g)                    0.14              3.500000  7395.650217   
 Fat (g)                      7.40            164.444444  7395.650217   
 Sugar (g)                    5.24            104.800000  7395.650217   
 
                         RMSE        MAE  
 Calories (kcal)    85.997966  41.478333  
 Protein (g)        85.997966  41.478333  
 Carbohydrates (g)  85.997966  41.478333  
 Fiber (g)          85.997966  41.478333  
 Fat (g)            85.997966  41.478333  
 Sugar (g)          85.997966  41.478333  ,
                    Absolute Error  Percentage Error (%)          MSE  \
 Calories (kcal)            149.76            149.760000  3772.861433   
 Protein (g)                 12.33     

### Metrics for LLaMA and Gemini

#### LLaMA
| Metric              | Calories (kcal) | Protein (g) | Carbohydrates (g) | Fiber (g) | Fat (g) | Sugar (g) |
|---------------------|-----------------|-------------|--------------------|-----------|---------|-----------|
| **Absolute Error**  | 209.52          | 17.76       | 8.81              | 0.14      | 7.40    | 5.24      |
| **Percentage Error (%)** | 209.52      | 253.71      | 58.73             | 3.50      | 164.44  | 104.80    |
| **MSE**             | 7395.65         | 7395.65     | 7395.65           | 7395.65   | 7395.65 | 7395.65   |
| **RMSE**            | 85.99           | 85.99       | 85.99             | 85.99     | 85.99   | 85.99     |
| **MAE**             | 41.48           | 41.48       | 41.48             | 41.48     | 41.48   | 41.48     |

---

#### Gemini
| Metric              | Calories (kcal) | Protein (g) | Carbohydrates (g) | Fiber (g) | Fat (g) | Sugar (g) |
|---------------------|-----------------|-------------|--------------------|-----------|---------|-----------|
| **Absolute Error**  | 149.76          | 12.33       | 6.24              | 1.00      | 4.14    | 0.07      |
| **Percentage Error (%)** | 149.76      | 176.14      | 41.60             | 25.00     | 92.00   | 1.40      |
| **MSE**             | 3772.86         | 3772.86     | 3772.86           | 3772.86   | 3772.86 | 3772.86   |
| **RMSE**            | 61.42           | 61.42       | 61.42             | 61.42     | 61.42   | 61.42     |
| **MAE**             | 28.92           | 28.92       | 28.92             | 28.92     | 28.92   | 28.92     |

---

### Key Insights:
- **LLaMA** has larger errors, particularly in **calories** and **fat**, with very high percentage errors across most metrics.
- **Gemini** performs better overall, with notably low error for **sugar** and more reasonable values for other nutrients.
- The **Root Mean Square Error (RMSE)** indicates that Gemini predictions are closer to real-world values compared to LLaMA.